### PR TITLE
feat: latest and earliest ByOffset UDFs to capture N values

### PR DESCRIPTION
**Pull request by sarwarbhuiyan**
### Description 
Extend latest_by_offset by accepting a parameter n which would then give you the latest n entries by a key as an array. This is also the same for earliest.

### Testing done 
Unit tests and functional tests updated

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by sarwarbhuiyan** [comment]
[clabot:check]

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java [hunk 1]
```diff
@@ -123,4 +155,54 @@ public T map(final Struct agg) {
...
+        merged.addAll(aggOne);
+        merged.addAll(aggTwo);
+        final int currentSize = merged.size();
+        Collections.sort(merged, structComparator);
+        if (currentSize <= topNSize) {
+          return merged;
+        } else {
+          return merged.subList(0, topNSize);
+        }
+      }
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java:199]: Can pre-allocate the array and simplify the code:

```suggestion
      @Override
      public List<Struct> merge(final List<Struct> aggOne, final List<Struct> aggTwo) {
        final List<Struct> merged = new ArrayList<>(aggOne.size() + aggTwo.size());
        merged.addAll(aggOne);
        merged.addAll(aggTwo);
        merged.sort(structComparator);
        return merged.subList(0, Math.min(topNSize, merged.size()));
      }
```

Plus, I noticed there's not test to cover the code path where the merged array is smaller than `topNSize`. Worth adding that.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestStructComparator.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+package io.confluent.ksql.function.udaf.offset;
+
+import static io.confluent.ksql.function.udaf.KudafByOffsetUtils.SEQ_FIELD;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import org.apache.kafka.connect.data.Struct;
+
+public class EarliestStructComparator implements Comparator<Struct>, Serializable {
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestStructComparator.java:24]: This class is duplicate `LatestStructComparator`, i.e. the code is the same!  And the function below duplicates the exist `KudafByOffsetUtils.compareStructs`.

Can I suggest deleting these two new comparator files, and instead replacing the `compareStructs` method in `KudafByOffsetUtils` with a comparator, e.g. 

```java
class  KudafByOffsetUtils {
  ...
  public static final Comparator<Struct> INTERMEDIATE_STRUCT_COMPARATOR = (struct1, struct2) -> {
    // Deal with overflow - we assume if one is positive and the other negative then the sequence
    // has overflowed - in which case the latest is the one with the smallest sequence
    final long sequence1 = struct1.getInt64(SEQ_FIELD);
    final long sequence2 = struct2.getInt64(SEQ_FIELD);
    if (sequence1 < 0 && sequence2 >= 0) {
      return 1;
    } else if (sequence2 < 0 && sequence1 >= 0) {
      return -1;
    } else {
      return Long.compare(sequence1, sequence2);
    }
  };
  ...
}
```

And then use this everywhere, including replacing the current use of `compareStructs` with `INTERMEDIATE_STRUCT_COMPARATOR.compare`.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/LatestStructComparator.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+package io.confluent.ksql.function.udaf.offset;
+
+import static io.confluent.ksql.function.udaf.KudafByOffsetUtils.SEQ_FIELD;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import org.apache.kafka.connect.data.Struct;
+
+public class LatestStructComparator implements Comparator<Struct>, Serializable {
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/LatestStructComparator.java:24]: As above, delete this.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/LatestByOffset.java [hunk 4]
```diff
@@ -40,43 +45,75 @@
...
       aggregateSchema = "STRUCT<SEQ BIGINT, VAL INT>")
   public static Udaf<Integer, Struct, Integer> latestInteger() {
     return latest(STRUCT_INTEGER);
   }
+  
+  @UdafFactory(description = "return the latest value of an integer column",
+      aggregateSchema = "ARRAY<STRUCT<SEQ BIGINT, VAL INT>>")
+  public static Udaf<Integer, List<Struct>, List<Integer>> latestInteger(final int topNSize) {
+    return latestN(STRUCT_INTEGER, topNSize);
+  }
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/LatestByOffset.java:67]: Same changes to the new functions in this file, as were done to the ones in `EarliestByOffset`, i.e. plural method names and param names changes to `latestN`.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java [hunk 5]
```diff
@@ -123,4 +155,54 @@ public T map(final Struct agg) {
       }
     };
   }
+
+
+  @UdafFactory(description = "Earliest N by offset")
+  static <T> Udaf<T, List<Struct>, List<T>> earliestN(final Schema structSchema,
+      final int topNSize) {
+    return new Udaf<T, List<Struct>, List<T>>() {
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java:163]: We'd want to validate `topNSize` (please add unit test for this too):

```suggestion
  static <T> Udaf<T, List<Struct>, List<T>> earliestN(
      final Schema structSchema,
      final int earliestN
    ) {
    
    if (earliestN <= 0) {
      throw new KsqlFunctionException("earliestN must be 1 or greater");
    }
    
    return new Udaf<T, List<Struct>, List<T>>() {
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/LatestByOffset.java [hunk 6]
```diff
@@ -124,5 +161,50 @@ public T map(final Struct agg) {
...
+        }
+      }
+
+      @Override
+      @SuppressWarnings("unchecked")
+      public List<T> map(final List<Struct> agg) {
+        return (List<T>) agg.stream().map(s -> s.get(VAL_FIELD)).collect(Collectors.toList());
+      }
+    };
+  }
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/LatestByOffset.java:208]: Please make corresponding changes to this code, as were suggested for the `EarliestByOffset` code.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java [hunk 7]
```diff
@@ -55,6 +70,29 @@ public void shouldComputeEarliestInteger() {
...
+  }
+  
+  @Test
+  public void shouldComputeEarliest2Integers() {
+    // Given:
+    final Udaf<Integer, List<Struct>, List<Integer>> udaf = EarliestByOffset.earliestInteger(2);
+
+    // When:
+    List<Struct> res = udaf
+        .aggregate(123, new ArrayList<>(Arrays.asList(EarliestByOffset.createStruct(STRUCT_INTEGER, 321))));
```
Comment (big-andy-coates) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java:82]: FYI: Just `Arrays.asList` needed. No need for the `new ArrayList<>(` bit.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java [hunk 8]
```diff
@@ -55,6 +70,29 @@ public void shouldComputeEarliestInteger() {
...
+    assertThat(res.get(1).get(VAL_FIELD), is(123));
+    
+    List<Struct> res2 = udaf
+        .aggregate(543, res);
+    assertThat(res2.size(), is(2));
+    assertThat(res2.get(0).get(VAL_FIELD), is(321));
+    assertThat(res2.get(1).get(VAL_FIELD), is(123));
+    
+    
+
```
Comment (big-andy-coates) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java:95]: Probably better structured as two tests:

```suggestion
@Test
  public void shouldCaptureValuesUpToN() {
    // Given:
    final Udaf<Integer, List<Struct>, List<Integer>> udaf = EarliestByOffset.earliestInteger(2);

    // When:
    List<Struct> res0 = udaf.aggregate(321, new ArrayList<>());
    List<Struct> res1 = udaf.aggregate(123, res0);

    assertThat(res1, hasSize(2));
    assertThat(res1.get(0).get(VAL_FIELD), is(321));
    assertThat(res1.get(1).get(VAL_FIELD), is(123));
  }

  @Test
  public void shouldIgnoreValuesPastN() {
    // Given:
    final Udaf<Integer, List<Struct>, List<Integer>> udaf = EarliestByOffset.earliestInteger(2);

    final List<Struct> aggregate = Arrays.asList(
        EarliestByOffset.createStruct(STRUCT_INTEGER, 10),
        EarliestByOffset.createStruct(STRUCT_INTEGER, 3)
    );

    // When:
    final List<Struct> result = udaf.aggregate(2, aggregate);

    assertThat(result, hasSize(2));
    assertThat(result.get(0).get(VAL_FIELD), is(10));
    assertThat(result.get(1).get(VAL_FIELD), is(3));
  }```

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java [hunk 9]
```diff
@@ -73,6 +111,28 @@ public void shouldMerge() {
...
+    List<Struct> agg2 = new ArrayList<>(Arrays.asList(struct3, struct4));
+    
+    // When:
+    List<Struct> merged1 = udaf.merge(agg1, agg2);
+    List<Struct> merged2 = udaf.merge(agg2, agg1);
+
+    // Then:
+    assertThat(merged1, is(agg1));
+    assertThat(merged2, is(agg1));
+  }
```
Comment (big-andy-coates) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java:135]: A better test would be if the structs in each parameter had interleaved sequence numbers, as this ensures the sort is happening:

```suggestion
  @Test
  public void shouldMergeIntegers() {
    // Given:
    final Udaf<Integer, List<Struct>, List<Integer>> udaf = EarliestByOffset.earliestInteger(2);

    final Struct struct1 = EarliestByOffset.createStruct(STRUCT_INTEGER, 123);
    final Struct struct2 = EarliestByOffset.createStruct(STRUCT_INTEGER, 321);
    final Struct struct3 = EarliestByOffset.createStruct(STRUCT_INTEGER, 543);
    final Struct struct4 = EarliestByOffset.createStruct(STRUCT_INTEGER, 654);
    final List<Struct> agg1 = new ArrayList<>(Arrays.asList(struct1, struct4));
    final List<Struct> agg2 = new ArrayList<>(Arrays.asList(struct2, struct3));
    
    // When:
    final List<Struct> merged1 = udaf.merge(agg1, agg2);
    final List<Struct> merged2 = udaf.merge(agg2, agg1);

    // Then:
    assertThat(merged1, contains(struct1, struct2));
    assertThat(merged2, contains(struct1, struct2));
  }
```

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java [hunk 10]
```diff
@@ -108,6 +194,26 @@ public void shouldComputeEarliestLong() {
     // Then:
     assertThat(res.getInt64(VAL_FIELD), is(321L));
   }
+  
+  @Test
+  public void shouldComputeEarliest2Longs() {
```
Comment (big-andy-coates) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java:199]:
```suggestion
  public void shouldComputeEarliestLongs() {
```

Same for other new methods.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java [hunk 11]
```diff
@@ -108,6 +194,26 @@ public void shouldComputeEarliestLong() {
...
+    // Given:
+    final Udaf<Long, List<Struct>, List<Long>> udaf = EarliestByOffset.earliestLong(2);
```
Comment (big-andy-coates) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java:201]: Consider changing the value of `earlestN`, so not all tests use `2`.  Alternate between 1, 2 & 3, maybe?

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/LatestByOffsetUdafTest.java [hunk 12]
```diff
@@ -45,19 +45,48 @@ public void shouldInitialize() {
     assertThat(init, is(notNullValue()));
   }
 
+  @Test
```
Comment (big-andy-coates) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/LatestByOffsetUdafTest.java:48]: Apply same changes to this file as were made to the `EarliestByOffsetTest` file.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java (hunk 5 above)
Comment (sarwarbhuiyan) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java:163]: Good point. Should there be an upper limit too?

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java [hunk 13]
```diff
@@ -123,4 +155,54 @@ public T map(final Struct agg) {
       }
     };
   }
+
+
+  @UdafFactory(description = "Earliest N by offset")
```
[the remark this answers is not in the document]
Comment (sarwarbhuiyan) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java:160]: will remove. it was in the original methods for earliest() and latest(). will remove those too.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java (hunk 1 above)
Comment (sarwarbhuiyan) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java:199]: added

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestStructComparator.java (hunk 2 above)
Comment (sarwarbhuiyan) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestStructComparator.java:24]: done

**Comment by sarwarbhuiyan** [comment]
I'm experiencing something weird. I force pushed my branch to the existing branch in sarwarbhuiyan/ksql but it's not showing up in the PR correctly other than above saying I force pushed the sarwarbhuiyan:ksql-latest-n-by-offset 14 hours ago.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java [hunk 14]
```diff
@@ -23,39 +23,110 @@
...
+    List<Struct> init = udaf.initialize();
+
+    // Then:
+    assertThat(init, is(notNullValue()));
+  }
+  
+  @Test
+  public void shouldThrowExceptionForInvalidN() {
+    try {
+      final Udaf<Integer, List<Struct>, List<Integer>> udaf = EarliestByOffset
```
Comment (big-andy-coates) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetUdafTest.java:69]: Fix spot bugs issue:

```suggestion
      EarliestByOffset
```

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/LatestByOffsetUdafTest.java [hunk 15]
```diff
@@ -96,57 +209,146 @@ public void shouldMergeWithOverflow() {
...
+
+    // When:
+    List<Struct> res = udaf
+        .aggregate(123L, Lists.newArrayList(LatestByOffset.createStruct(STRUCT_LONG, 321L)));
+    
+    List<Struct> res2 = udaf
+        .aggregate(543L, res);
+    
+    List<Struct> res3 = udaf
+        .aggregate(654L, res);
```
Comment (big-andy-coates) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/LatestByOffsetUdafTest.java:263]: Fix spot bugs issue:

Either asset the contents of `res` and `res2` or remove them as variables.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/LatestByOffsetUdafTest.java [hunk 16]
```diff
@@ -45,19 +51,83 @@ public void shouldInitialize() {
...
+    List<Struct> init = udaf.initialize();
+
+    // Then:
+    assertThat(init, is(notNullValue()));
+  }
+  
+  @Test
+  public void shouldThrowExceptionForInvalidN() {
+    try {
+      final Udaf<Integer, List<Struct>, List<Integer>> udaf = LatestByOffset
```
Comment (big-andy-coates) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/offset/LatestByOffsetUdafTest.java:69]: Fix spot bugs error:

```suggestion
      LatestByOffset
```

**Review by big-andy-coates — COMMENTED** [review]
Build failing with spot bugs issues:

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java [hunk 17]
```diff
@@ -38,8 +42,9 @@
     author = KsqlConstants.CONFLUENT_AUTHOR
 )
 public final class EarliestByOffset {
+
   static final String DESCRIPTION =
-      "This function returns the oldest value for the column, computed by offset.";
+      "This function returns the oldest N value for the column, computed by offset.";
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java:47]: nit:

```suggestion
      "This function returns the oldest N values for the column, computed by offset.";
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/LatestByOffset.java [hunk 18]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2019 Confluent Inc.
+ * Copyright 2020 Confluent Inc.
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/LatestByOffset.java:2]: please don't update existing copyright dates

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java [hunk 19]
```diff
@@ -52,30 +57,60 @@ private EarliestByOffset() {
...
+  @UdafFactory(description = "return the earliest N values of a double column",
+      aggregateSchema = "ARRAY<STRUCT<SEQ BIGINT, VAL DOUBLE>>")
+  public static Udaf<Double, List<Struct>, List<Double>> earliestDoubles(final int earliestN) {
+    return earliestN(STRUCT_DOUBLE, earliestN);
+  }
 
   @UdafFactory(description = "return the earliest value of a boolean column",
       aggregateSchema = "STRUCT<SEQ BIGINT, VAL BOOLEAN>")
-  public static Udaf<Boolean, Struct, Boolean> earliestBoolean() {
+  public static Udaf<Boolean, Struct, Boolean> earliestBooleans() {
```
Comment (big-andy-coates) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java:92]: nit: 

```suggestion
  public static Udaf<Boolean, Struct, Boolean> earliestBoolean() {
```

**Review by big-andy-coates — COMMENTED** [review]
LGTM, 'cept for a few nits and I've just realized there's no documentation update. Would you mind updating `LATEST_BY_OFFSET` and `EARLIEST_BY_OFFSET` in https://github.com/confluentinc/ksql/blob/master/docs/developer-guide/ksqldb-reference/aggregate-functions.md so that it covers the new optional parameter.

**Comment by big-andy-coates** [comment]
Hey @sarwarbhuiyan,

The last commit titled 'fixed ksqldb-functional-tests' fixes a doc issue, (LATEST_BY_OFFSET can't be applied to a table), but also includes generated JavaDocs in the `ksqldb-functional-tests/doc` dir. Can you please delete this dir.

**Review by big-andy-coates — APPROVED** [review]
LGTM - thanks @sarwarbhuiyan !!!

**Pull request merged**